Support Azure Blob Storage in COPY TO/COPY FROM

BaurzhanSakhariev opened the pull request:
Closes https://github.com/crate/crate/issues/13852

- [x] Docs
- [x] More unit tests and use either Azurite imitation via HttpServer or use testcontainers
- [x] Validation (including unknown props validation, something we don't have even on master)
- [x] properly load opendal_java  (probably as runtime dependency, not test dependency) https://github.com/apache/opendal/tree/main/bindings/java#classified-library 
- [x] Figure out why Itest reports leaked threads (happens both with local instance or testcontainer, ie a real issue)
- [x] SAS token

BaurzhanSakhariev commented:
test_copy_to_and_copy_from_azure_blob_storage itself passes but causes thread leaks (in my TODO) - CI not green.

mfussenegger commented:
> Introduced Scheme class to store all OpenDAL specific constants, so that we can incrementally add other provideres (GCS and others).

I'm not sure that's a good idea.
Despite using OpenDAL it would be good to keep the different copy implementations decoupled. Adding a component that bundles all of the settings together runs counter to that.

The `server` module should ideally stay entirely unaware of any specific implementations and have no knowledge about which concrete scheme and properties are supported.


> Per-provider settings grouping. We currently don't do smth like WITH (s3.property) but instead put all settings flattened and mention in docs (for example, protocol setting is only s3 specific).

This would be a breaking change and I don't really see why it is an improvement? In any case, it shouldn't be changed in this PR.

mfussenegger commented on plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileInput.java:
```diff
@@ -0,0 +1,150 @@
+ * Outgoing URI-s are then used by other components and sent back to this component,
+ * so outgoing format (expandURI) implicitly dictates incoming URI-s format (getStream).
+ */
+public class AzureFileInput implements FileInput {
```
Maybe something for a follow up where we migrate s3 to OpenDAL too, but I was hoping that the per-scheme implementation becomes much smaller by using it.

E.g. We could have a `copy` module which defines a `OpenDALInput` which can be parameterized with the supported scheme/properties and anything that can be specific to a backend. It looks like only the URI building needs some custom logic?

BaurzhanSakhariev replied:
> It looks like only the URI building needs some custom logic?

Yes

So basically you mean moving out as much as code as possible to the common module and make OpenDAL dependant plugins work only with non-common logic?

It makes thanks, will do (as a follow if it's fine), want to finish working MVP first as this refactoring won't affect user facing functionality.

Just to be sure:

Given that Azure plugin is the only one using OpenDAL currently, total amount of code won't change.

> per-scheme implementation becomes much smaller 

will be true only once we migrate S3 or/and add new plugins for other providers

mfussenegger replied:
> (as a follow if it's fine).

yes, probably better anyway to do it in another step - together with migrating s3. Always helps to have 2-3 implementations to validate an abstraction.

mfussenegger commented on plugins/crate-copy-azure/src/test/java/io/crate/copy/azure/AzureCopyIntegrationTest.java:
```diff
@@ -0,0 +1,124 @@
+
+    @Before
+    public void setup() {
+        azureContainer = new GenericContainer<>("mcr.microsoft.com/azure-storage/azurite:3.31.0")
```
If we start using containers we'd need to update the dev docs to mention the docker/podman(?) dependency.

Is the existing azure stub server that we have not good enough?

BaurzhanSakhariev commented:
>This would be a breaking change and I don't really see why it is an improvement? In any case, it shouldn't be changed in this PR.

I didn't want to change it, just highlights

> I suggest to keep flat settings structure in order not to introduce breaking user facing changes.

So, yes, will keep it flat

BaurzhanSakhariev commented:
Will use a dedicated lib, problem described in https://github.com/orgs/crate/projects/21?pane=issue&itemId=78426356 persists even if I use single `AsyncExecutor`. 

See also https://github.com/apache/opendal/discussions/5096

BaurzhanSakhariev commented on server/src/testFixtures/java/org/elasticsearch/repositories/azure/AzureHttpHandler.java:
```diff
@@ -177,13 +186,22 @@ public void handle(final HttpExchange exchange) throws IOException {
                     list.append("<Blob><Name>").append(blobPath).append("</Name>");
                     list.append("<Properties><Content-Length>").append(blob.getValue().length()).append(
                         "</Content-Length>");
+                    List<String> dateHeader = exchange.getRequestHeaders().get("X-ms-date");
+                    if (dateHeader != null) {
+                        // COPY/OpenDAL requires this header.
+                        // CREATE REPOSITORY/dedicated client fails on encountering header in such format.
+                        // TODO: Migrate CREATE REPOSITORY on Azure to use OpenDAL.
+                        list.append("<Last-Modified>").append(dateHeader.get(0)).append("</Last-Modified>");
+                    }
```
Some settings are common but Repository uses dedicated library and it doesn't play well with OpenDAL. I added todo to maybe use OpenDAL in es-azure plugin as well but for now I added dedicated `AzureBlobStorageSettings`.

mfussenegger commented on plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureBlobStorageSettings.java:
```diff
@@ -0,0 +1,75 @@
+    static final Setting<String> CONTAINER_SETTING = Setting.simpleString("container", Property.NodeScope);
+    static final Setting<String> ENDPOINT_SETTING = Setting.simpleString("endpoint", Property.NodeScope);
+
+    public static Map<Setting<String>, Boolean> settings() {
```
I think his should have some description about what the boolean value signals

mfussenegger reviewed (commented):
> Got rid of testcontainers, moved AzureHttpHandler to the server so that different plugins can see it. Had to tweak some headers a little bit so that it works with OpenDAL and doesn't lead to timeouts/bad request errors.

I'd tend to add a azure-testing module that we can use as dependency where needed instead of adding it to server.

---

overall approach seems okay to me

BaurzhanSakhariev commented on plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileInput.java:
```diff
@@ -0,0 +1,164 @@
+     */
+    @Override
+    public List<URI> expandUri() throws IOException {
+        try (Operator operator = operator()) {
```
Updated. Please note, that initially I wanted to close `FileInput` on `closeReader` but those 2 have different lifecycles:

We can have reader/stream closed but `currentInput` still open:
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java#L326-L327

Hence, closing `FileInput` in `reset` before we nullify it. 
I don't catch exceptions on `currentInput.close()`: my thinking is that we should propagate `reset` exceptions further?

Added no-op default implementation of `close` to the `FileInput` as other implementations don't have anything to close.

Unit testing was a bit cumbersome and all this boilerplate with having `operator()` method and calling instead of using `this.operator` in `expandURI` is caused by not wanting to create `Operator` outside of `AzureFileInput`.

BaurzhanSakhariev replied:
Pushed another commit to make `FileOutput` `AutoCloseable`.

Simply added new field `FileOutput` to the `RowWriter`. 

Tried to also remove `OutputStream` field from the `RowWriter` and acquire it based on `FileOutput` but passing `acquireOutputStream` params (or Function to apply) looked more like introducing more mess so I just added a new field.

BaurzhanSakhariev commented on plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/SharedAsyncExecutor.java:
```diff
@@ -0,0 +1,62 @@
+        return asyncExecutor;
+    }
+
+    @Override
+    protected void doStart() {
+        // No-op.
+    }
+
+    @Override
+    protected void doStop() {
+        // No-op.
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+        asyncExecutor.close();
+    }
```
Not sure I understood the problem:
`....closed, started...` means we re-created all objects and each new "started" has it's own `AsyncExecutor`?

Double close shouldn't be a problem (even if possible). `AsyncExecutor` is a `NativeObject` with `close` looking like
```
if (disposed.compareAndSet(false, true)) {
     disposeInternal(nativeHandle);
}
```
ie second close is a no-op.

What's a suggested scenario?

1. COPY...
2. cluster().fullRestart();
3. COPY...
4. Check that we don't have lingering OpenDAL threads with names having pattern `Thread-x`?

BaurzhanSakhariev commented:
> I'd tend to add a azure-testing module that we can use as dependency where needed instead of adding it to server.


This would be a module of a single file, I started adding it to `crate-libs-shared` to avoid having a module with a single file.

Then I realized that `AzureHttpHandler` requires some files from server:
```
import org.elasticsearch.common.bytes.BytesArray;
import org.elasticsearch.common.bytes.BytesReference;
import org.elasticsearch.common.io.Streams;
```
A new module would have to depend on server. 

Plugins which are already depend on `server` and now depending on `azure-testing` (or any another module depending on server and "hosting" `AzureHttpHandler`) would have a circular dependency?

mfussenegger commented:
> Plugins which are already depend on server and now depending on azure-testing (or any another module depending on server and "hosting" AzureHttpHandler) would have a circular dependency?


It's not circular:

```
    server◄────────┬
      ▲            │
      │            │
  azure-testing    │
      ▲            │
      │            │
   azure-copy──────┘
```

It definitely shouldn't go into `crate-libs-shared` - the idea is to contain the dependencies to where they are used & needed, not spread them throughout all modules.

BaurzhanSakhariev commented:
Tried adding a new module in in https://github.com/crate/crate/pull/16600/, maven install is successful but IDEA shows an error:

`Annotation processing is not supported for module cycles.`

Checking it further

BaurzhanSakhariev commented:
Added a new module in https://github.com/crate/crate/pull/16600

BaurzhanSakhariev commented:
Hi @mfussenegger could you please take another look?

Addressed most comments (except restart thing, seehttps://github.com/crate/crate/pull/16468#discussion_r1752304296)

mfussenegger reviewed (commented):
lgtm so far

BaurzhanSakhariev commented:
Hi @seut, could you take a look? 

Got one lgtm but added docs now. 

Already [highlighted](https://github.com/crate/crate/pull/16468#issuecomment-2340566745) places that might needs extra attention.

seut commented on plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureBlobStorageSettings.java:
```diff
@@ -0,0 +1,94 @@
+     * Rejects unknown settings.
+     * @param commonCopySettings is either common COPY TO or common COPY FROM settings.
+     */
+    public static Map<String, String> openDALConfig(Settings settings, List<String> commonCopySettings) {
```
Also, why do we need to pass in here `commonCopySettings` if they are addressable in a static way anyhow?

BaurzhanSakhariev replied:
>can be done where it is needed

I introduced this because we need it in 2 places (FileInput/FileOutput). Will split it up to avoid confusion.

> why do we need to pass in here commonCopySettings
it's wither COPY_TO, or COPY_FROM, depending on entry point (FileInput/FileOutput). But yeah, doesn't make sense to pass static, will do smth else.

BaurzhanSakhariev replied:
I added azure->opendal names rewriting in https://github.com/crate/crate/pull/16468/commits/b8d7ffe260efa4486f05ef894c1a4bfaf2fb529c

Feel like we actually need 1 method to build OpenDAL config.
Also, will be able to get rid of one test (https://github.com/crate/crate/pull/16468#discussion_r1757073329)

seut commented on plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileInput.java:
```diff
@@ -0,0 +1,147 @@
+
+    @VisibleForTesting
+    Operator operator() {
+        return AsyncOperator.of(NAME, config, sharedAsyncExecutor.asyncExecutor()).blocking();
```
why is this not returning the existing `this.operator`?

seut replied:
Seems like this is only used inside tests, so please move this out of production code. But I still don't understand why this is needed at all.

BaurzhanSakhariev replied:
Oh, that's a wrong thing, some residuals related to mocking `operator.list` in `test_expand_uri` and not wanting to create Operator outside of `AzureFileInput`. 

I guess I have to pass `Operator` to make it mockable (ie create it outside or pass Function), will do.

seut commented on plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileInput.java:
```diff
@@ -0,0 +1,147 @@
+     */
+    @Override
+    public List<URI> expandUri() throws IOException {
+        var operator = operator(); // Only for tests, otherwise could use this.operator
```
The comment doesn't tell me anything valuable. Why is any test related stuff here inside production code?

BaurzhanSakhariev replied:
Removed, see https://github.com/crate/crate/pull/16468#discussion_r1757240817

seut commented on plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileInput.java:
```diff
@@ -0,0 +1,147 @@
+    private final Map<String, String> config;
+    private final URI uri;
+    private final GlobPredicate uriPredicate;
+    private final String preGlobPath;
```
```suggestion
    @Nullable
    private final String preGlobPath;
```

seut commented on plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileInput.java:
```diff
@@ -0,0 +1,147 @@
+            return List.of(uri);
+        }
+        List<URI> uris = new ArrayList<>();
+        List<Entry> entries = operator.list(preGlobPath);
```
Is it save to pass a nullable value to `operator.list()`?

BaurzhanSakhariev replied:
YES, because we never get there if preGlobPath is null. 

We shortcut on `if (isGlobbed() == false)`  which is false on NULL. I will add an assertion.

seut commented on docs/sql/statements/copy-from.rst:
```diff
@@ -287,6 +287,37 @@ Using the ``s3://`` scheme automatically sets the `shared`_ to true.
+
+.. code-block:: text
+
+    azblob://dir1/dir2/file1.json
```
A valid URI without any host component must use 3 slashes.
```suggestion
    azblob:///dir1/dir2/file1.json
```

BaurzhanSakhariev replied:
I aligned with s3 https://cratedb.com/docs/crate/reference/en/latest/sql/statements/copy-from.html#s3

Will update and also will check if it helps to get rid of `AzureFileOutput.resourcePath` and use `uri.getPath()`.

seut commented on docs/sql/statements/copy-to.rst:
```diff
@@ -281,6 +281,35 @@ See `AWS Java Documentation`_.
+
+.. code-block:: text
+
+    azblob://dir1/dir2/file1.json
```
Same here, this isn't a valid URI.
```suggestion
    azblob:///dir1/dir2/file1.json
```

seut commented on plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileOutput.java:
```diff
@@ -0,0 +1,73 @@
+        operator.close();
+    }
+
+    /**
+     * Extracts resource path from a user provided URI (azblob:://path/to/dir).
+     */
+    public static String resourcePath(URI uri) {
+        // We cannot use uri.getPath() since it treats first directory as a host.
+        return uri.toString().replace("azblob:/", "");
+    }
```
Not needed when using and documenting the use of a valid URI (`azblob:///path/dir` <- 3 slashes after the schema name)
```suggestion
```

seut commented on plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileInput.java:
```diff
@@ -0,0 +1,147 @@
+     */
+    @Override
+    public InputStream getStream(URI uri) throws IOException {
+        return operator.createInputStream(uri.toString());
```
```suggestion
        return operator.createInputStream(uri.getPath());
```

seut commented on plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileInput.java:
```diff
@@ -0,0 +1,147 @@
+        // Pre-glob path operates with user-provided URI as GLOB pattern reflects user facing COPY FROM uri format.
+        this.preGlobPath = toPreGlobPath(uri);
+
+        String resourcePath = resourcePath(uri);
+        this.uri = URI.create(resourcePath);
```
```suggestion
        this.uri = uri;
```

seut commented on plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileInput.java:
```diff
@@ -0,0 +1,147 @@
+    public static String toPreGlobPath(URI uri) {
+        Matcher hasGlobMatcher = HAS_GLOBS_PATTERN.matcher(uri.toString());
+        if (hasGlobMatcher.matches()) {
+            return resourcePath(URI.create(hasGlobMatcher.group(1)));
```
I guess the `URI.create()` roundtrip is purely for URI string parsing validation. If already validate it wouldn't make any sense.
```suggestion
            return URI.create(hasGlobMatcher.group(1)).toString();
```

seut commented on plugins/crate-copy-azure/src/test/java/io/crate/copy/azure/AzureCopyIntegrationTest.java:
```diff
@@ -0,0 +1,144 @@
+public class AzureCopyIntegrationTest extends IntegTestCase {
+
+    private static final String CONTAINER_NAME = "test";
+    private static final String AZURITE_ACCOUNT = "devstoreaccount1";
```
```suggestion
    private static final String AZURE_ACCOUNT = "devstoreaccount1";
```

seut commented on plugins/crate-copy-azure/src/test/java/io/crate/copy/azure/AzureCopyIntegrationTest.java:
```diff
@@ -0,0 +1,144 @@
+
+    private static final String CONTAINER_NAME = "test";
+    private static final String AZURITE_ACCOUNT = "devstoreaccount1";
+    private static final String AZURITE_KEY =
```
```suggestion
    private static final String AZURE_KEY =
```

seut commented on plugins/crate-copy-azure/src/test/java/io/crate/copy/azure/AzureCopyIntegrationTest.java:
```diff
@@ -0,0 +1,144 @@
+    private String endpoint;
+
+    private HttpServer httpServer;
+    private AzureHttpHandler handler;
```
Can be a local var, not used anywhere else.

BaurzhanSakhariev replied:
httpServer is stopped on `tearDown`, made `AzureHttpHandler` local.

seut commented on plugins/crate-copy-azure/src/test/java/io/crate/copy/azure/AzureCopyIntegrationTest.java:
```diff
@@ -0,0 +1,144 @@
+    private static final String CONTAINER_NAME = "test";
+    private static final String AZURITE_ACCOUNT = "devstoreaccount1";
+    private static final String AZURITE_KEY =
+            "Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==";
```
To clarify that this is just a dummy.
```suggestion
            "dummy-key;
```

seut commented on plugins/crate-copy-azure/src/test/java/io/crate/copy/azure/AzureFileInputTest.java:
```diff
@@ -0,0 +1,100 @@
+    public void test_to_pre_glob_path() {
+        var uris =
+            List.of(
+                "azblob://dir1/prefix*/*.json",
```
as said, all these uris must use 3 slashes after the schema.

seut commented on plugins/crate-copy-azure/src/test/java/io/crate/copy/azure/AzureFileInputTest.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    @Test
+    public void test_reject_unknown_setting() throws Exception {
+        Settings settings = Settings.builder().put("dummy", "dummy").build();
+
+        assertThatThrownBy(() -> new AzureFileInput(mock(SharedAsyncExecutor.class), URI.create("azblob://dir1/dir2/*"), settings))
+            .isExactlyInstanceOf(IllegalArgumentException.class)
+            .hasMessage("Setting 'dummy' is not supported");
+    }
```
Could be moved into a dedicated `AzureBlobStorageSettingsTest`

seut commented on plugins/crate-copy-azure/src/test/java/io/crate/copy/azure/AzureFileOutputTest.java:
```diff
@@ -0,0 +1,41 @@
+public class AzureFileOutputTest {
+
+    @Test
+    public void test_reject_unknown_setting() throws Exception {
+        Settings settings = Settings.builder().put("dummy", "dummy").build();
+
+        assertThatThrownBy(() -> new AzureFileOutput(mock(SharedAsyncExecutor.class), settings))
+            .isExactlyInstanceOf(IllegalArgumentException.class)
+            .hasMessage("Setting 'dummy' is not supported");
+    }
```
Can be removed if the logic is already tested by a dedicated `AzureBlobStorageSettingsTest`

BaurzhanSakhariev replied:
I will remove `openDALConfig` and call `validate` in 2 places (https://github.com/crate/crate/pull/16468#discussion_r1757059322), we would need 2 tests as Input/Output as validating in one place doesn't guarantee that we have validation call in another.

BaurzhanSakhariev commented on docs/sql/statements/copy-to.rst:
```diff
@@ -281,6 +281,35 @@ See `AWS Java Documentation`_.
+    TO DIRECTORY 'azblob://dir1/dir2'
+    WITH (
+        container = 'container',
+        account_name = 'account_name',
+        account_key = 'account_key',
```
I aligned with OpenDAL parameter names (https://docs.rs/opendal/latest/opendal/services/struct.Azblob.html).

I will align user facing params with Azure CREATE REPOSITORY and will transform internally to OpenDAL compatible names.

BaurzhanSakhariev commented on docs/sql/statements/copy-to.rst:
```diff
@@ -281,6 +281,35 @@ See `AWS Java Documentation`_.
+        container = 'container',
+        account_name = 'account_name',
+        account_key = 'account_key',
+        endpoint = 'endpoint'
```
I opened https://github.com/crate/crate/pull/16607, still WIP (plan was to get at least auth method merged) and incrementally add token

BaurzhanSakhariev replied:
Added changes from another PR here

BaurzhanSakhariev commented:
>Can't we put the AzureStorageSettings class somewhere into a shared lib? I really don't like the duplication of all Azure settings (registry + validation).

I'm not sure all settings listed in `AzureStorageSettings` (and used by a dedicated client, coming from  <artifactId>azure-storage</artifactId>) are supported by OpenDAL which lists only few params in docs (https://docs.rs/opendal/latest/opendal/services/struct.Azblob.html).

Also, some settings have irrelevant defaults - `CONTAINER_SETTING` has default "crate-snapshots", there is no required/not-required concept. 

I will try it out by putting it to a new module from https://github.com/crate/crate/pull/16600. Will rename it from `azure-testing` to `azure-common`. I got quote some problems with IDEA/broken project structure after changing modules, if it's fine, I will try it after wrapping up this PR

seut commented:
> I'm not sure all settings listed in `AzureStorageSettings` (and used by a dedicated client, coming from azure-storage) are supported by OpenDAL which lists only few params in docs (https://docs.rs/opendal/latest/opendal/services/struct.Azblob.html).

This could be handled by different lists of supported settings. But it would align on the names, types and maybe validation of such settings. Given that you already used different names for the same settings, it may make sense to align it.

BaurzhanSakhariev commented on docs/sql/statements/copy-to.rst:
```diff
@@ -355,11 +384,57 @@ the copy operation to complete. If set to ``false`` the request
+``container``
+'''''''''''''
+
+Used for :ref:`azblob <sql-copy-to-azblob>` scheme only.
```
Done in https://github.com/crate/crate/pull/16468/commits/e536c47728a8b2775f23d98be93d17f0f92822c7

proddata commented on docs/sql/statements/copy-from.rst:
```diff
@@ -287,6 +289,38 @@ Using the ``s3://`` scheme automatically sets the `shared`_ to true.
+    COPY source
+    TO DIRECTORY 'azblob:///dir1/dir2'
+    WITH (
+        container = 'container',
```
Why is container separated and not part of the URI? Doesn't seem quite consistent with how this is handled on Azure.

BaurzhanSakhariev replied:
Went in this direction (all in `WITH`) as we already separate 2 params and keeping all granular was not a big effort, and could be also more convenient for users. 

For example, `az storage container create` accepts endpoint AND account/key as a separate params (even if account is a part of endpoint). 

We can also support URI with account/container, smth like
 
 `https://myaccount.blob.core.windows.net/mycontainer/myvirtualdirectory/myblob`, plan is to discuss it later and maybe actually keep all-in-WITH and instead adjust S3.

proddata replied:
I think it is less convenient and also eg doesn't follow what other solutions accept.

Container is always required. Everything else in `WITH` could be optional.

Also if we adjust s3 import typically paths are provided as `s3://<bucket>/<blob>` (eg in the AWS Console)

proddata commented on docs/sql/statements/copy-from.rst:
```diff
@@ -287,6 +289,38 @@ Using the ``s3://`` scheme automatically sets the `shared`_ to true.
+
+    azblob:///dir1/dir2/file1.json
+
+All other parameters must be provided in the ``WITH`` clause.
```
Should they not all be optional?

BaurzhanSakhariev replied:
`container` and `endpoint` are required

proddata replied:
Isn't endpoint almost always  blob.core.windows.net for azure blob storage?

BaurzhanSakhariev replied:
URI in CrateDB must start with scheme name (s3/file/azblob), that's kinda tech requirement to load proper plugin.

If we support full endpoint, it won't be `COPY TO https:...`, it will be `COPY TO azblob:... ` and we would have to introduce dedicated endpoint param anyway.

proddata replied:
Why? Parsing the URI to determine whether it’s a full endpoint or just a container name is straightforward, and this actually supports the idea of including the container name in the URI. The entire purpose of a URI is to uniquely identify a resource, which isn’t possible without specifying the container.

https://duckdb.org/docs/extensions/azure.html#for-azure-blob-storage
